TFNetworkLayer.py

```diff
@@ -1107,53 +1111,218 @@ def batch_norm(self, data,
           with self.var_creation_scope():
             beta = self.add_param(tf.Variable(
-              initial_value=tf.zeros(data.get_bc_spatial_batch_shape()),
+              initial_value=tf.zeros(shape),
               name="%s_%s_beta" % (self.name, data.name),
               trainable=True))
+        else:
+          beta = tf.reshape(beta, shape)
+
+      if fused:
+        # It's possible to use fused bn operation
+        return self._fused_batch_norm(
+          data=data, axes=axes,
+          use_shift=use_shift, use_std=use_std,
+          momentum=momentum, epsilon=epsilon,
+          sample_mean=sample_mean, sample_variance=sample_variance,
+          gamma=gamma, beta=beta,
+          masked_time=masked_time)
+      if masked_time:
+        # Default case
+        x = input_data.get_placeholder_flattened(keep_dims=True)
+        mean, variance = tf.nn.moments(x, axes=[0], keep_dims=True)
+        mean = tf.reshape(mean, shape=variables_shape)
+        variance = tf.reshape(variance, shape=variables_shape)
+      else:
+        axes = input_data.get_axes_from_description(axes)
+        assert input_data.feature_dim_axis in axes
+        assert input_data.batch_dim_axis not in axes  # Not supported
+        assert input_data.time_dim_axis not in axes  # Not supported
+        x = input_data.placeholder
+        axes_norm = axes
+        axes_stat = list(set(input_data.get_axes()) - set(axes_norm))
+        mean, variance = tf.nn.moments(x, axes=axes_stat, keep_dims=True)
+      # Use exponential moving average of batch mean.
+      # Note: We could also use cumulative moving average. Our Theano implementation does that for inference.
+      if isinstance(sample_mean, tf.Variable):
+        sample_mean = tf.assign_add(sample_mean, (mean - sample_mean) * momentum)
+      if isinstance(sample_variance, tf.Variable):
+        sample_variance = tf.assign_add(sample_variance, (variance - sample_variance) * momentum)
+      # If train or if force_sample, use default use_sample=0.0, otherwise use_sample=1.0.
+      use_sample = 1.0 + tf.cast(tf.logical_or(self.network.train_flag, force_sample), tf.float32) * (use_sample - 1.0)
+      mean = (1. - use_sample) * mean + use_sample * sample_mean
+      variance = (1. - use_sample) * variance + use_sample * sample_variance
+      bn = (input_data.placeholder - mean) * tf.rsqrt(variance + epsilon)
+      if use_std:
+        bn *= gamma
+      if use_shift:
         bn += beta
+      if data.is_batch_feature_major and not masked_time:
+        # Transpose back to NCHW
+        perm = data.get_axes()
+        perm = perm[:1] + perm[-1:] + perm[1:-1]
+        bn = tf.transpose(bn, perm=perm)
       return bn
 
+  def _fused_batch_norm(self, data,
+                        axes,
+                        use_shift, use_std,
+                        momentum, epsilon,
+                        sample_mean, sample_variance,
+                        gamma, beta,
+                        masked_time):
```

Comment:
As I explained already, don't make it a separate function. You have lot of duplicated code here.